expression/util.go

```diff
@@ -146,30 +166,48 @@ func setExprColumnInOperand(expr Expression) Expression {
 // ColumnSubstitute substitutes the columns in filter to expressions in select fields.
 // e.g. select * from (select b as a from t) k where a < 10 => select * from (select b as a from t where b < 10) k.
 func ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression) Expression {
+	_, resExpr := ColumnSubstituteImpl(expr, schema, newExprs)
+	return resExpr
+}
+
+// ColumnSubstituteImpl tries to substitute column expr using newExprs,
+// the newFunctionInternal is only called if its child is substituted
+func ColumnSubstituteImpl(expr Expression, schema *Schema, newExprs []Expression) (bool, Expression) {
 	switch v := expr.(type) {
 	case *Column:
 		id := schema.ColumnIndex(v)
 		if id == -1 {
-			return v
+			return false, v
 		}
 		newExpr := newExprs[id]
 		if v.InOperand {
 			newExpr = setExprColumnInOperand(newExpr)
 		}
-		return newExpr
+		return true, newExpr
 	case *ScalarFunction:
 		if v.FuncName.L == ast.Cast {
 			newFunc := v.Clone().(*ScalarFunction)
-			newFunc.GetArgs()[0] = ColumnSubstitute(newFunc.GetArgs()[0], schema, newExprs)
-			return newFunc
+			_, newFunc.GetArgs()[0] = ColumnSubstituteImpl(newFunc.GetArgs()[0], schema, newExprs)
+			return true, newFunc
+		}
+		refExprArr := cowExprRef{v.GetArgs(), nil}
+		substituted := false
+		for idx, arg := range v.GetArgs() {
+			changed, newFuncExpr := ColumnSubstituteImpl(arg, schema, newExprs)
+			if substituted {
```

Comment:
This seems complex, how about a better wrap:

```
func Set(i, change, val) {
     if new != nil {
        new[i] = val
        return
     }

     if !change {
        return
     }

     new = make()
     copy(i)
     new[i] = val
}
```

And use `refExprArr.Set(idx, changed, newFunExpr)` here?

Reply:
And better with comments what `cowExprRef` does, otherwise, you'll find the code tricky after 3 months